src/Features/LanguageServer/Protocol/Handler/SignatureHelp/SignatureHelpHandler.cs

```diff
@@ -0,0 +1,166 @@
+                        if (clientCapabilities?.HasVisualStudioLspCapability() == true)
+                        {
+                            sigInfo = new LSP.VSSignatureInformation
+                            {
+                                ColorizedLabel = GetSignatureClassifiedText(item)
+                            };
+                        }
+                        else
+                        {
+                            sigInfo = new LSP.SignatureInformation();
+                        }
+
+                        sigInfo.Label = GetSignatureText(item);
+                        sigInfo.Documentation = new LSP.MarkupContent { Kind = LSP.MarkupKind.PlainText, Value = item.DocumentationFactory(cancellationToken).GetFullText() };
+                        sigInfo.Parameters = item.Parameters.Select(p => new LSP.ParameterInformation
+                        {
+                            Label = p.Name,
+                            Documentation = new LSP.MarkupContent { Kind = LSP.MarkupKind.PlainText, Value = p.DocumentationFactory(cancellationToken).GetFullText() }
+                        }).ToArray();
+                        sigInfos.Add(sigInfo);
+                    }
+
+                    var sigHelp = new LSP.SignatureHelp
+                    {
+                        ActiveSignature = GetActiveSignature(items),
+                        ActiveParameter = items.ArgumentIndex,
+                        Signatures = sigInfos.ToArray()
+                    };
+
+                    return sigHelp;
+                }
+            }
+
+            return new LSP.SignatureHelp();
+        }
+
+
+        private static int GetActiveSignature(SignatureHelpItems items)
+        {
+            if (items.SelectedItemIndex.HasValue)
+            {
+                return items.SelectedItemIndex.Value;
+            }
+
+            // From Roslyn's doc comments for SelectedItemIndex:
+            //   If this is null, then the controller will pick the first item that has enough arguments
+            //   to be viable based on what argument position the user is currently inside of.
+            // However, the LSP spec expects the language server to make this decision.
+            // So implement the logic of picking a signature that has enough arguments here.
+
+            var matchingSignature = items.Items.FirstOrDefault(sig => sig.Parameters.Length > items.ArgumentIndex);
+            return matchingSignature != null ? items.Items.IndexOf(matchingSignature) : 0;
+        }
+
+        /// <summary>
+        /// The <see cref="SignatureHelpItem"/> contains a prefix, parameters seperated by a
+        /// separator and a suffix. Parameters themselves have a prefix, display and suffix.
+        /// Concatenate them all to get the text.
+        /// </summary>
+        private string GetSignatureText(SignatureHelpItem item)
+        {
+            var sb = new StringBuilder();
```

Comment:
I think we should have work item to create better markup content for signature help and completion and etc.

we should have all info needed to create better markup content, we just didn't spend time to do so?

Reply:
https://github.com/dotnet/roslyn/projects/45#card-20033878